packages/flutter_tools/lib/src/commands/upgrade.dart

```diff
@@ -222,10 +225,36 @@ class UpgradeCommandRunner {
     }
   }
 
+  /// Checks if the Flutter git repository is tracking a "standard remote".
+  ///
+  /// Using `flutter upgrade` is not supported from a non-standard remote. A git
+  /// remote should have the same url as [_flutterGit] to be considered as a
+  /// "standard" remote.
+  ///
+  /// Exits tool if the tracking remote is not standard.
+  void verifyStandardRemote(FlutterVersion localVersion) {
+    if (localVersion.repositoryUrl != _flutterGit) {
```

Comment:
If any user sets the tracking repository as `https://github.com/flutter/flutter`(no ".git" suffix) and having `FLUTTER_GIT_URL` unset, this will fail. How do I handle this case?

Reply:
Good catch, and that's annoying. You could write a utility function and pass both `localVersion.repositoryUrl` and `_flutterGit` through it before comparing. Something like:

```
String stripDotGit(String s) {
  RegExp pattern = RegExp(r'(.*)(\.git)$');
  final RegExpMatch match = pattern.firstMatch(s);
  if (match == null) {
    return s;
  }
  return match.group(1);
}
```

As with any regex, we'd want to unit test this :)